### PR TITLE
Fix learning phase info being left out in multi-input models

### PR DESCRIPTION
**Pull request by gokceneraslan**
Batch size support for histograms were introduced in #6065. This, unfortunately, breaks multi-input or multi-output models as it assumes that `val_data` has a fixed length of 3 (or 4 if self.mode.uses_learning_phase it True). This commit changes the behavior by slicing everything in val_data, excluding the learning_phase, because the length of `val_data` is determined by the number of inputs and outputs.

### CODE REVIEW

**Comment by gokceneraslan** [comment]
I think this must be included in the next release, because all multi-input/output model using TB callback and learning phase fail with `You must feed a value for placeholder tensor 'dense_1/keras_learning_phase' with dtype bool` error now.

**Comment by fchollet** [comment]
Please add a unit test.

**Comment by gokceneraslan** [comment]
Done.

**Comment by gokceneraslan** [comment]
Speaking of tests, it would be extra nice if you can merge #7024 as well.

File: tests/keras/test_callbacks.py [hunk 1]
```diff
@@ -355,30 +352,22 @@ def data_generator_graph(train):
...
+              callbacks=cbks, epochs=3)
 
     # fit with validation data and accuracy
-    model.fit(X_train, y_train, batch_size=batch_size,
-              validation_data=(X_test, y_test), callbacks=cbks, epochs=2)
-
-    # fit generator with validation data
-    model.fit_generator(data_generator(True), len(X_train), epochs=2,
-                        validation_data=(X_test, y_test),
-                        callbacks=cbks)
```
Comment (fchollet) [inline tests/keras/test_callbacks.py]: You are removing previous test cases. Add a new test rather than editing an existing one.

Comment (gokceneraslan) [inline tests/keras/test_callbacks.py]: That one was already redundant. It's same as fourth one below. Or you mean I shouldn't replace single input models with multi-input models and add a new test instead? I thought multi-input would already cover single-input cases.

Comment (gokceneraslan) [inline tests/keras/test_callbacks.py]: Convnet TB cases are single input already by the way.

Comment (fchollet) [inline tests/keras/test_callbacks.py]: Test is not redundant. Please fix.

**Comment by gokceneraslan** [comment]
I added single input/output tests back. There are 8 tests in total now, namely:

- with and without generators,
- with and without the validation data.
- single and multi input/output models

This makes 8 tests. You'll see that there are still some test removed. This is because there were some duplicate tests. See below for clarification:

same

**Pull request merged**